### PR TITLE
feat(ado-improvements-1): Refactor reporter errors

### PR DESCRIPTION
**Pull request by DaveTryon**
#### Details

As called out in #950, we throw an `Error` if a scan fails, then we log the `Error` twice. To a user, it looks like our tool has crashed, when in reality it's an accessibility error that the user should fix. This PR does the following to improve this:

1. Some implementations of `ProgressReporter.failRun` used to throw an Error to force the program to fail. The new code merely sets a flag in `ProgressReporter.failRun`, then a new API, `ProgressReporter.didScanSucceed`, serves as an indicator of whether or not the scan succeeded.
2. Code that called `ProgressReporter.failRun` with a message is now responsible for logging the error message via `Logger.logError` or `Logger.trackException`, depending on the circumstance. With this change order of the `ProgressReporter` objects no longer matters, so I removed the associated comment.
3. `Scanner.scan` now returns a `Promise<boolean>` instead of a `Promise<void>`. The resolved value of the promise indicates user action. A successful value (true) means that no corrective action is needed by the user. A failed value (false) means that the user need to take some corrective action.
4. The application now has 3 supported return codes
  - A value of 0 (`ExitCode.ScanCompletedNoUserActionIsNeeded`) means that the scan completed successfully and no corrective user action is needed.
  - A value of 1 (`ExitCode.ScanCompletedUserActionIsNeeded`) means that the scan completed successfully, but corrective user action is needed.
  - A value of 2 (`ExitCode.ScanFailedToComplete`) means that the scan failed to complete. This will always be accompanied by a log of the Error that led to the failure.

I also wordsmithed a couple of error messages while I was updating the code and associated tests. A couple of linting issues were also flagged, so I tidied them up.

##### Motivation

Make progress toward #950

##### Before and after

<details>
<summary>
Output before this change--baseline file doesn't match the scan results. Notice the duplicated stack dumps and the fact that the error message is lost as the message of the stack dump
</summary>

```
##[group]Installing runtime dependencies
[1/4] 🔍  Resolving packages...
warning Lockfile has incorrect entry for "apify-shared@^0.5.0". Ignoring it.
warning Lockfile has incorrect entry for "socket.io@^2.3.0". Ignoring it.
warning Lockfile has incorrect entry for "underscore@1.12.1". Ignoring it.
warning Lockfile has incorrect entry for "apify-shared@^0.6.0". Ignoring it.
warning Lockfile has incorrect entry for "apify-shared@^0.1.45". Ignoring it.
warning Lockfile has incorrect entry for "css-what@^5.0.0". Ignoring it.
warning Lockfile has incorrect entry for "http-signature@~1.2.0". Ignoring it.
warning Lockfile has incorrect entry for "css-what@2.1". Ignoring it.
warning Lockfile has incorrect entry for "nth-check@~1.0.1". Ignoring it.
warning Lockfile has incorrect entry for "axios@^0.21.1". Ignoring it.
warning Lockfile has incorrect entry for "marked@^2.0.0". Ignoring it.
warning accessibility-insights-scan > apify > socket.io > debug@4.1.1: Debug versions >=3.2.0 <3.2.7 || >=4 <4.3.1 have a low-severity ReDos regression when used in a Node.js environment. It is recommended you upgrade to 3.2.7 or 4.3.1. (https://github.com/visionmedia/debug/issues/797)
warning accessibility-insights-scan > apify > socket.io > socket.io-parser > debug@4.1.1: Debug versions >=3.2.0 <3.2.7 || >=4 <4.3.1 have a low-severity ReDos regression when used in a Node.js environment. It is recommended you upgrade to 3.2.7 or 4.3.1. (https://github.com/visionmedia/debug/issues/797)
warning accessibility-insights-scan > apify > socket.io > engine.io > debug@4.1.1: Debug versions >=3.2.0 <3.2.7 || >=4 <4.3.1 have a low-severity ReDos regression when used in a Node.js environment. It is recommended you upgrade to 3.2.7 or 4.3.1. (https://github.com/visionmedia/debug/issues/797)
[2/4] 🚚  Fetching packages...
[3/4] 🔗  Linking dependencies...
[4/4] 🔨  Building fresh packages...
##[endgroup]
##vso[task.debug]agent.TempDirectory=undefined
##vso[task.debug]agent.workFolder=undefined
##vso[task.debug]loading inputs and endpoints
##vso[task.debug]loading INPUT_OUTPUTDIR
##vso[task.debug]loading INPUT_SITEDIR
##vso[task.debug]loading INPUT_SCANURLRELATIVEPATH
##vso[task.debug]loading INPUT_MAXURLS
##vso[task.debug]loading INPUT_SCANTIMEOUT
##vso[task.debug]loading INPUT_SINGLEWORKER
##vso[task.debug]loading INPUT_URL
##vso[task.debug]loading INPUT_BASELINEFILE
##vso[task.debug]loaded 8
##vso[task.debug]Agent.ProxyUrl=undefined
##vso[task.debug]Agent.CAInfo=undefined
##vso[task.debug]Agent.ClientCert=undefined
##vso[task.debug]Agent.SkipCertValidation=undefined
##vso[task.debug]scanTimeout=90000
##vso[task.debug]url=https://markreay.github.io/AU/before.html
##vso[task.debug]inputFile=undefined
##vso[task.debug]inputUrls=undefined
##vso[task.debug]discoveryPatterns=undefined
##vso[task.debug]outputDir=_accessibility-reports
##vso[task.debug]maxUrls=100
##vso[task.debug]chromePath=undefined
##vso[task.debug]url=https://markreay.github.io/AU/before.html
##vso[task.debug]singleWorker=true
##vso[task.debug]baselineFile=./test.baseline
##[debug]Starting accessibility scanning of URL https://markreay.github.io/AU/before.html
##[debug]Chrome app executable: system default
##[debug] PuppeteerCrawler:AutoscaledPool:Snapshotter: Setting max memory of this run to 4096 MB. Use the APIFY_MEMORY_MBYTES environment variable to override it.
##[debug] PuppeteerCrawler:AutoscaledPool: state {"currentConcurrency":0,"desiredConcurrency":1,"systemStatus":{"isSystemIdle":true,"memInfo":{"isOverloaded":false,"limitRatio":0.2,"actualRatio":null},"eventLoopInfo":{"isOverloaded":false,"limitRatio":0.4,"actualRatio":null},"cpuInfo":{"isOverloaded":false,"limitRatio":0.4,"actualRatio":null},"clientInfo":{"isOverloaded":false,"limitRatio":0.3,"actualRatio":null}}}
##[debug] Launching Puppeteer {"args":["--disable-dev-shm-usage","--no-sandbox","--disable-setuid-sandbox","--js-flags=--max-old-space-size=8192","--no-sandbox","--user-agent=Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/86.0.4240.75 Safari/537.36"],"defaultViewport":{"width":1920,"height":1080,"deviceScaleFactor":1},"timeout":15000,"headless":true}
Processing page https://markreay.github.io/AU/before.html
Discovered 4 links on page https://markreay.github.io/AU/before.html
Found 4 accessibility issues on page https://markreay.github.io/AU/before.html
##[debug] PuppeteerCrawler: All the requests from request list and/or request queue have been processed, the crawler will shut down.
##[debug] PuppeteerCrawler: Final request statistics: {"requestAvgFailedDurationMillis":null,"requestAvgFinishedDurationMillis":2914,"requestsFinishedPerMinute":20,"requestsFailedPerMinute":0,"requestTotalDurationMillis":2914,"requestsTotal":1,"crawlerRuntimeMillis":3033,"requestsFinished":1,"requestsFailed":0,"retryHistogram":[1]}
##vso[task.debug]outputDir=_accessibility-reports
Scan report saved successfully as /Users/dtryon/repos/ada/action/main/packages/ado-extension/dist/pkg/_accessibility-reports/index.html
##[debug]Found 22 new violations compared to the baseline.
##[debug]Found 0 cases where a previously baselined violation was fixed.
##[debug]Saved new baseline file at /Users/dtryon/repos/ada/action/main/packages/ado-extension/dist/pkg/_accessibility-reports/test.baseline
##[debug]To update the baseline with these changes, either rerun with --updateBaseline or copy the updated baseline file to /Users/dtryon/repos/ada/action/main/packages/ado-extension/dist/pkg/test.baseline
##vso[task.debug]baselineFile=./test.baseline
### Results from Current Run
### ![Accessibility Insights](https://accessibilityinsights.io/img/a11yinsights-blue.svg) Accessibility Insights Action

**22 failure instances**
* (4) **color-contrast**:  Ensures the contrast between foreground and background colors meets WCAG 2 AA contrast ratio thresholds
* (1) **html-has-lang**:  Ensures every HTML document has a lang attribute
* (4) **image-alt**:  Ensures \<img> elements have alternate text or a role of none or presentation
* (13) **label**:  Ensures every form element has a label

**Baseline not detected**
To update the baseline with these changes, copy the updated baseline file to [scan arguments](undefined). See [baselining docs](https://github.com/microsoft/accessibility-insights-action/blob/main/docs/ado-extension-usage.md#using-a-baseline-file) for more.

See all failures and scan details by downloading the report from [run artifacts](undefined)

---
#### Scan summary
**URLs**: 1 with failures, 0 passed, 0 not scannable
**Rules**: 22 with failures, 11 passed, 37 not applicable

---
This scan used [axe-core 4.3.2](https://github.com/dequelabs/axe-core/releases/tag/v4.3.2) with Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/86.0.4240.75 Safari/537.36.
##vso[task.debug]failOnAccessibilityError=undefined
##vso[task.debug]baselineFile=./test.baseline
##[error][Exception]VError: An error occurred while scanning website page https://markreay.github.io/AU/before.html: Failed: The baseline file does not match scan results. If this is a PR, check the PR comments.
    at Logger.trackExceptionAny (/Users/dtryon/repos/ada/action/main/packages/ado-extension/dist/pkg/index.js:134706:29)
    at Scanner.<anonymous> (/Users/dtryon/repos/ada/action/main/packages/ado-extension/dist/pkg/index.js:135826:29)
    at Generator.throw (<anonymous>)
    at rejected (/Users/dtryon/repos/ada/action/main/packages/ado-extension/dist/pkg/index.js:135755:65) {
  jse_shortmsg: 'An error occurred while scanning website page https://markreay.github.io/AU/before.html',
  jse_cause: Error: Failed: The baseline file does not match scan results. If this is a PR, check the PR comments.
      at WorkflowEnforcer.failIfBaselineNeedsUpdating (/Users/dtryon/repos/ada/action/main/packages/ado-extension/dist/pkg/index.js:130124:19)
      at WorkflowEnforcer.<anonymous> (/Users/dtryon/repos/ada/action/main/packages/ado-extension/dist/pkg/index.js:130108:18)
      at Generator.next (<anonymous>)
      at /Users/dtryon/repos/ada/action/main/packages/ado-extension/dist/pkg/index.js:130085:71
      at new Promise (<anonymous>)
      at __webpack_modules__../src/progress-reporter/enforcement/workflow-enforcer.ts.__awaiter (/Users/dtryon/repos/ada/action/main/packages/ado-extension/dist/pkg/index.js:130081:12)
      at WorkflowEnforcer.completeRun (/Users/dtryon/repos/ada/action/main/packages/ado-extension/dist/pkg/index.js:130106:16)
      at /Users/dtryon/repos/ada/action/main/packages/ado-extension/dist/pkg/index.js:135391:45
      at AllProgressReporter.<anonymous> (/Users/dtryon/repos/ada/action/main/packages/ado-extension/dist/pkg/index.js:135409:27)
      at Generator.next (<anonymous>)
      at fulfilled (/Users/dtryon/repos/ada/action/main/packages/ado-extension/dist/pkg/index.js:135366:58),
  jse_info: {}
}
Accessibility scanning of URL https://markreay.github.io/AU/before.html completed
##[debug]Exception thrown in extension:  Error: Error: Failed: The baseline file does not match scan results. If this is a PR, check the PR comments.
    at WorkflowEnforcer.failIfBaselineNeedsUpdating (/Users/dtryon/repos/ada/action/main/packages/ado-extension/dist/pkg/index.js:130124:19)
    at WorkflowEnforcer.<anonymous> (/Users/dtryon/repos/ada/action/main/packages/ado-extension/dist/pkg/index.js:130108:18)
    at Generator.next (<anonymous>)
    at /Users/dtryon/repos/ada/action/main/packages/ado-extension/dist/pkg/index.js:130085:71
    at new Promise (<anonymous>)
    at __webpack_modules__../src/progress-reporter/enforcement/workflow-enforcer.ts.__awaiter (/Users/dtryon/repos/ada/action/main/packages/ado-extension/dist/pkg/index.js:130081:12)
    at WorkflowEnforcer.completeRun (/Users/dtryon/repos/ada/action/main/packages/ado-extension/dist/pkg/index.js:130106:16)
    at /Users/dtryon/repos/ada/action/main/packages/ado-extension/dist/pkg/index.js:135391:45
    at AllProgressReporter.<anonymous> (/Users/dtryon/repos/ada/action/main/packages/ado-extension/dist/pkg/index.js:135409:27)
    at Generator.next (<anonymous>)
    at fulfilled (/Users/dtryon/repos/ada/action/main/packages/ado-extension/dist/pkg/index.js:135366:58)
    at WorkflowEnforcer.<anonymous> (/Users/dtryon/repos/ada/action/main/packages/ado-extension/dist/pkg/index.js:130114:19)
    at Generator.next (<anonymous>)
    at /Users/dtryon/repos/ada/action/main/packages/ado-extension/dist/pkg/index.js:130085:71
    at new Promise (<anonymous>)
    at __webpack_modules__../src/progress-reporter/enforcement/workflow-enforcer.ts.__awaiter (/Users/dtryon/repos/ada/action/main/packages/ado-extension/dist/pkg/index.js:130081:12)
    at WorkflowEnforcer.failRun (/Users/dtryon/repos/ada/action/main/packages/ado-extension/dist/pkg/index.js:130113:16)
    at /Users/dtryon/repos/ada/action/main/packages/ado-extension/dist/pkg/index.js:135400:41
    at AllProgressReporter.<anonymous> (/Users/dtryon/repos/ada/action/main/packages/ado-extension/dist/pkg/index.js:135409:27)
    at Generator.next (<anonymous>)
    at fulfilled (/Users/dtryon/repos/ada/action/main/packages/ado-extension/dist/pkg/index.js:135366:58)
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

</details>

<details>
<summary>
Output after this change--baseline file doesn't match the scan results. Notice that the stack dumps are removed and that the message is called out with an `##[error]` tag near the end of the output
</summary>

```
##[group]Installing runtime dependencies
[1/4] 🔍  Resolving packages...
warning Lockfile has incorrect entry for "apify-shared@^0.5.0". Ignoring it.
warning Lockfile has incorrect entry for "socket.io@^2.3.0". Ignoring it.
warning Lockfile has incorrect entry for "underscore@1.12.1". Ignoring it.
warning Lockfile has incorrect entry for "apify-shared@^0.6.0". Ignoring it.
warning Lockfile has incorrect entry for "apify-shared@^0.1.45". Ignoring it.
warning Lockfile has incorrect entry for "css-what@^5.0.0". Ignoring it.
warning Lockfile has incorrect entry for "http-signature@~1.2.0". Ignoring it.
warning Lockfile has incorrect entry for "css-what@2.1". Ignoring it.
warning Lockfile has incorrect entry for "nth-check@~1.0.1". Ignoring it.
warning accessibility-insights-scan > apify > socket.io > debug@4.1.1: Debug versions >=3.2.0 <3.2.7 || >=4 <4.3.1 have a low-severity ReDos regression when used in a Node.js environment. It is recommended you upgrade to 3.2.7 or 4.3.1. (https://github.com/visionmedia/debug/issues/797)
warning accessibility-insights-scan > apify > socket.io > socket.io-parser > debug@4.1.1: Debug versions >=3.2.0 <3.2.7 || >=4 <4.3.1 have a low-severity ReDos regression when used in a Node.js environment. It is recommended you upgrade to 3.2.7 or 4.3.1. (https://github.com/visionmedia/debug/issues/797)
warning accessibility-insights-scan > apify > socket.io > engine.io > debug@4.1.1: Debug versions >=3.2.0 <3.2.7 || >=4 <4.3.1 have a low-severity ReDos regression when used in a Node.js environment. It is recommended you upgrade to 3.2.7 or 4.3.1. (https://github.com/visionmedia/debug/issues/797)
warning Lockfile has incorrect entry for "axios@^0.21.1". Ignoring it.
warning Lockfile has incorrect entry for "marked@^2.0.0". Ignoring it.
[2/4] 🚚  Fetching packages...
[3/4] 🔗  Linking dependencies...
[4/4] 🔨  Building fresh packages...
##[endgroup]
##vso[task.debug]agent.TempDirectory=undefined
##vso[task.debug]agent.workFolder=undefined
##vso[task.debug]loading inputs and endpoints
##vso[task.debug]loading INPUT_OUTPUTDIR
##vso[task.debug]loading INPUT_SITEDIR
##vso[task.debug]loading INPUT_SCANURLRELATIVEPATH
##vso[task.debug]loading INPUT_MAXURLS
##vso[task.debug]loading INPUT_SCANTIMEOUT
##vso[task.debug]loading INPUT_SINGLEWORKER
##vso[task.debug]loading INPUT_URL
##vso[task.debug]loading INPUT_BASELINEFILE
##vso[task.debug]loaded 8
##vso[task.debug]Agent.ProxyUrl=undefined
##vso[task.debug]Agent.CAInfo=undefined
##vso[task.debug]Agent.ClientCert=undefined
##vso[task.debug]Agent.SkipCertValidation=undefined
##vso[task.debug]scanTimeout=90000
##vso[task.debug]url=https://markreay.github.io/AU/before.html
##vso[task.debug]inputFile=undefined
##vso[task.debug]inputUrls=undefined
##vso[task.debug]discoveryPatterns=undefined
##vso[task.debug]outputDir=_accessibility-reports
##vso[task.debug]maxUrls=100
##vso[task.debug]chromePath=undefined
##vso[task.debug]url=https://markreay.github.io/AU/before.html
##vso[task.debug]singleWorker=true
##vso[task.debug]baselineFile=./test.baseline
##[debug]Starting accessibility scanning of URL https://markreay.github.io/AU/before.html
##[debug]Chrome app executable: system default
##[debug] PuppeteerCrawler:AutoscaledPool:Snapshotter: Setting max memory of this run to 4096 MB. Use the APIFY_MEMORY_MBYTES environment variable to override it.
##[debug] PuppeteerCrawler:AutoscaledPool: state {"currentConcurrency":0,"desiredConcurrency":1,"systemStatus":{"isSystemIdle":true,"memInfo":{"isOverloaded":false,"limitRatio":0.2,"actualRatio":null},"eventLoopInfo":{"isOverloaded":false,"limitRatio":0.4,"actualRatio":null},"cpuInfo":{"isOverloaded":false,"limitRatio":0.4,"actualRatio":null},"clientInfo":{"isOverloaded":false,"limitRatio":0.3,"actualRatio":null}}}
##[debug] Launching Puppeteer {"args":["--disable-dev-shm-usage","--no-sandbox","--disable-setuid-sandbox","--js-flags=--max-old-space-size=8192","--no-sandbox","--user-agent=Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/86.0.4240.75 Safari/537.36"],"defaultViewport":{"width":1920,"height":1080,"deviceScaleFactor":1},"timeout":15000,"headless":true}
Processing page https://markreay.github.io/AU/before.html
Discovered 4 links on page https://markreay.github.io/AU/before.html
Found 4 accessibility issues on page https://markreay.github.io/AU/before.html
##[debug] PuppeteerCrawler: All the requests from request list and/or request queue have been processed, the crawler will shut down.
##[debug] PuppeteerCrawler: Final request statistics: {"requestAvgFailedDurationMillis":null,"requestAvgFinishedDurationMillis":2824,"requestsFinishedPerMinute":20,"requestsFailedPerMinute":0,"requestTotalDurationMillis":2824,"requestsTotal":1,"crawlerRuntimeMillis":2972,"requestsFinished":1,"requestsFailed":0,"retryHistogram":[1]}
##vso[task.debug]outputDir=_accessibility-reports
Scan report saved successfully as /Users/dtryon/repos/ada/action/MyFork/packages/ado-extension/dist/pkg/_accessibility-reports/index.html
##[debug]Found 22 new violations compared to the baseline.
##[debug]Found 0 cases where a previously baselined violation was fixed.
##[debug]Saved new baseline file at /Users/dtryon/repos/ada/action/MyFork/packages/ado-extension/dist/pkg/_accessibility-reports/test.baseline
##[debug]To update the baseline with these changes, either rerun with --updateBaseline or copy the updated baseline file to /Users/dtryon/repos/ada/action/MyFork/packages/ado-extension/dist/pkg/test.baseline
##vso[task.debug]baselineFile=./test.baseline
### Results from Current Run
### ![Accessibility Insights](https://accessibilityinsights.io/img/a11yinsights-blue.svg) Accessibility Insights Action

**22 failure instances**
* (4) **color-contrast**:  Ensures the contrast between foreground and background colors meets WCAG 2 AA contrast ratio thresholds
* (1) **html-has-lang**:  Ensures every HTML document has a lang attribute
* (4) **image-alt**:  Ensures \<img> elements have alternate text or a role of none or presentation
* (13) **label**:  Ensures every form element has a label

**Baseline not detected**
To update the baseline with these changes, copy the updated baseline file to [scan arguments](undefined). See [baselining docs](https://github.com/microsoft/accessibility-insights-action/blob/main/docs/ado-extension-usage.md#using-a-baseline-file) for more.

See all failures and scan details by downloading the report from [run artifacts](undefined)

---
#### Scan summary
**URLs**: 1 with failures, 0 passed, 0 not scannable
**Rules**: 22 with failures, 11 passed, 37 not applicable

---
This scan used [axe-core 4.3.2](https://github.com/dequelabs/axe-core/releases/tag/v4.3.2) with Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/86.0.4240.75 Safari/537.36.
##vso[task.debug]failOnAccessibilityError=undefined
##vso[task.debug]baselineFile=./test.baseline
##[error]The baseline file does not match scan results.
Accessibility scanning of URL https://markreay.github.io/AU/before.html completed
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

</details>

#### Pull request checklist

- [x] Addresses an existing issue: Progress against #950
- [x] Added relevant unit test for your changes. (`yarn test`)
- [x] Verified code coverage for the changes made. Check coverage report at: `<rootDir>/test-results/unit/coverage`
- [x] Ran precheckin (`yarn precheckin`)

### CODE REVIEW

File: packages/shared/src/scanner/scanner.ts [hunk 1]
```diff
@@ -44,16 +42,17 @@ export class Scanner {
         @inject(BaselineFileUpdater) private readonly baselineFileUpdater: BaselineFileUpdater,
     ) {}
 
-    public async scan(): Promise<void> {
+    // Return indicates the scan status. It returns true if no corrective user action is needed.
+    public async scan(): Promise<boolean> {
```
Comment (karanbirsingh) [inline packages/shared/src/scanner/scanner.ts:46]: I wonder if creating a type for the return value here might be more readable. Something like 'UserActionRequired' and it could alias to a boolean.

File: packages/shared/src/progress-reporter/all-progress-reporter.ts [hunk 2]
```diff
@@ -26,8 +26,20 @@ export class AllProgressReporter extends ProgressReporter {
...
-    public async failRun(message: string): Promise<void> {
-        await this.execute((r) => r.failRun(message));
+    public async failRun(): Promise<void> {
+        await this.execute((r) => r.failRun());
+    }
+
+    public async didScanSucceed(): Promise<boolean> {
+        let allReportersSucceeded = true;
+
+        await this.execute(async (r) => {
```
Comment (karanbirsingh) [inline packages/shared/src/progress-reporter/all-progress-reporter.ts:36]: A different way to write this code is to return false early as soon as any progress reporter fails. But that only works if it's OK to skip running `didScanSucceed` for a given reporter. For example, maybe a progress reporter has some logging inside `didScanSucceed` that we can't omit. To me, I don't think progress reporters should expect that method must be called. But it would be helpful for me to clarify that assumption either way.

If the other approach (returning early) is functionally equivalent, I have a slight preference for it, but not too strong.

Comment (DaveTryon) [reply packages/shared/src/progress-reporter/all-progress-reporter.ts:36]: It functionally shouldn't matter which `ProgressReporter` reports the failure, and there's no intent for the `didScanSucceed` method to do anything interesting, but the way the `execute` method is defined, it only supports methods that return `Promise<void>`. I didn't want to create a parallel version of `execute`, so I settled for doing more work than is strictly necessary

File: packages/shared/src/scanner/scanner.ts (hunk 1 above)
Comment (DaveTryon) [reply packages/shared/src/scanner/scanner.ts]: It's a bit wordy, but I went with `ScanSucceededWithNoRequiredUserAction`. That allowed me to leave the true/false logic unchanged

File: packages/ado-extension/src/ado-extension.ts [hunk 3]
```diff
@@ -20,9 +21,9 @@ export function runScan() {
         await logger.setup();
 
         const scanner = container.get(Scanner);
-        await scanner.scan();
+        process.exit((await scanner.scan()) ? ExitCode.ScanCompletedNoUserActionIsNeeded : ExitCode.ScanCompletedUserActionIsNeeded);
     })().catch((error) => {
-        console.log('Exception thrown in extension: ', error);
-        process.exit(1);
+        console.log('##[error][Exception] Exception thrown in extension: ', error);
```
[the remark this answers is not in the document]
Comment (DaveTryon) [reply packages/ado-extension/src/ado-extension.ts:26]: That's correct. We get a Logger further upstream, but it a failure occurs before it's initialized, using console.log ensures that we can always produce the output

File: packages/gh-action/src/index.ts [hunk 4]
```diff
@@ -18,8 +19,8 @@ import { setupIocContainer } from './ioc/setup-ioc-container';
     await logger.setup();
 
     const scanner = container.get(Scanner);
-    await scanner.scan();
+    process.exit((await scanner.scan()) ? ExitCode.ScanCompletedNoUserActionIsNeeded : ExitCode.ScanCompletedUserActionIsNeeded);
 })().catch((error) => {
-    console.log('Exception thrown in action: ', error);
-    process.exit(1);
+    console.log('::error::[Exception] Exception thrown in extension: ', error);
```
Comment (DaveTryon) [inline packages/gh-action/src/index.ts:24]: This uses the action error syntax described at https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#setting-an-error-message. I'll create a separate PR that uses this through the `Logger` API

**Review by brocktaylor7 — APPROVED** [review]
